File: codechef/MSTEP.py
T = int(raw_input())

for it in range(T):
    N = int(raw_input())
    N2 = []
    for i in range(N):
        N2.append(map(int, raw_input().split()))
    
    mapping = [[0,0] for i in range(1, N**2 + 1)]
    for i in range(N):
        for j in range(N):
            mapping[N2[i][j] - 1] = [i, j]
    
    steps = 0
    curr = mapping[0]
    for i in range(1, N**2):
        dest = mapping[i]
        while(curr[0] != dest[0]):
            if curr[0] < dest[0]:
                curr[0] += 1
            else:
                curr[0] -= 1
            steps += 1
        while(curr[1] != dest[1]):
            if curr[1] < dest[1]:
                curr[1] += 1
            else:
                curr[1] -= 1
            steps += 1
    print steps

File: codechef/MSTEP.c
#include <stdio.h>
#include <stdlib.h>

#define inp getchar_unlocked

typedef long long int lld;

inline int input() {
    lld num = 0;
    char s = inp();
    while(s<'0'||s>'9') s = inp();
    while(s>='0'&&s<='9') num = (num<<3) + (num<<1) + (s-'0'), s = inp();
    return num;
}

typedef struct {
    int x, y;
} coord ;

int main() {
    int T, N, i, j, a, steps;
    
    coord curr, dest;
    coord *map;
    
    T = input();

    while(T--) {
        N  = input();
        map = (coord *) malloc(N*N * sizeof(coord));
        for(i=0; i<N; i++) {
            for(j=0; j<N; j++) {
                a = input();
                map[a-1].x = i;
                map[a-1].y = j;
            }
        }
               
        steps = 0;
        curr = map[0];
        for (i=1; i<N*N; i++) {
            dest = map[i];
            
            while(curr.x != dest.x) {
                if (curr.x < dest.x)
                    curr.x++;
                else
                    curr.x--;
                
                steps++;
            }
            
            while(curr.y != dest.y) {
                if (curr.y < dest.y) 
                    curr.y++;
                else
                    curr.y--;
                    
                steps++;
            }
            
        }
        free(map);
        printf("%d\n", steps);
    }
    return 0;
}
